### PR TITLE
add `rlang::check_installed()` to `assert_package()`

### PR DESCRIPTION
**Pull request by malcolmbarrett**
…and bump rlang requirement

# Prework

* [x] I understand and agree to the [code of conduct](https://ropensci.org/code-of-conduct/) and the [contributing guidelines](https://github.com/ropensci/targets/blob/main/CONTRIBUTING.md).
* [ ] I have already submitted a [discussion topic](https://github.com/ropensci/targets/discussions) or [issue](http://github.com/ropensci/targets/issues) to discuss my idea with the maintainer. 
* [x] Ignored above because this is so small 😇 
* [x] This pull request is not a [draft](https://github.blog/2019-02-14-introducing-draft-pull-requests).

# PR

This PR modifies `assert_package()` to use `rlang::check_installed()` which is a recent, very ergonomic package check that offers to install the package(s) for you. It's also vectorized. This did require a small bump in the required rlang version, as it's a new feature.

On my machine, for instance, I now get this message with `tar_watch()`:

```r
tar_watch()
ℹ The `bs4Dash` and `shinyWidgets` packages are required.
x Would you like to install them?

1: Yes
2: No
```

### CODE REVIEW

**Comment by wlandau** [comment]
I do like the menu, and I appreciate how much pain it takes away from the user. However, `check_installed()` throws an error of class `c("rlang_error", "error", "condition")` instead of the desired `c("condition_validate", "condition_targets", "error", "condition")`. Custom assertions like `assert_package()` generate custom error objects: 

https://github.com/ropensci/targets/blob/8b10d76b88954e75e7e1c2ba384f7426881e84eb/R/utils_condition.R#L98-L103

These custom error classes help `targets` and its tests distinguish among different kinds of errors and help keep the infrastructure working. If you can figure out how to combine the best of both worlds, I will accept the PR.

**Comment by wlandau** [comment]
Assuming we solve that, please add `check_installed` to the `@importFrom` statement in `R/tar_package.R`.

https://github.com/ropensci/targets/blob/8b10d76b88954e75e7e1c2ba384f7426881e84eb/R/tar_package.R#L23

**Comment by malcolmbarrett** [comment]
Ah, I see! For my own interest, can you point me to a spot where targets is checking such classed conditions? I have a growing interest in them after seeing some cool use cases.

Anyway, I did a simple catch of the error, replace it with `throw_validate()`, then imported `check_installed()`. Let me know how it looks!

File: R/utils_assert.R [hunk 1]
```diff
@@ -192,10 +192,13 @@ assert_nzchar <- function(x, msg = NULL) {
   }
 }
 
-assert_package <- function(package, msg = NULL) {
-  if (!requireNamespace(package, quietly = TRUE)) {
-    throw_validate(msg %||% paste("package", package, "not installed"))
-  }
+assert_package <- function(package) {
+  tryCatch(
+    check_installed(package),
```
Comment (wlandau) [inline R/utils_assert.R:197]: Would you namespace this call with `rlang::`? I know `check_installed()` is already in `@importFrom`, but I do both for the sake of defensive programming.

File: R/tar_package.R [hunk 2]
```diff
@@ -20,7 +20,8 @@
 #'   graph_from_data_frame igraph_opt igraph_options is_dag simplify topo_sort
 #'   V
 #' @importFrom R6 R6Class
-#' @importFrom rlang as_function as_quosure enquo quo quo_squash sym
+#' @importFrom rlang as_function as_quosure enquo quo quo_squash sym
```
Comment (wlandau) [inline R/tar_package.R:23]: Found a small lint here, just a trailing whitespace.

**Comment by wlandau** [comment]
I tested this out, and `check_installed()` makes it so easy to install missing packages without any interruption. I think the PR will be ready to merge after a couple of minor details are resolved (see my inline comments). Thanks for the contribution.

> For my own interest, can you point me to a spot where targets is checking such classed conditions? I have a growing interest in them after seeing some cool use cases.

I mostly use it for tests, as you may have seen with `expect_error(..., class = "condition_validate")`. It may be possible to define nice print methods for these kinds of errors, though I have not done so yet. On rare occasions, I also use it to divert control flow without having to worry about where I am in the call stack. This is useful in high-performance computing situations where we run targets locally until we find a target that requires a remote worker:

https://github.com/ropensci/targets/blob/1e4dfe184473ad0bc660564be409d44ddaf08f5e/R/class_prelocal.R#L54

https://github.com/ropensci/targets/blob/1e4dfe184473ad0bc660564be409d44ddaf08f5e/R/class_clustermq.R#L199-L205

File: R/utils_assert.R (hunk 1 above)
Comment (malcolmbarrett) [reply R/utils_assert.R:197]: Interesting, haven't seen that pattern before!

**Comment by malcolmbarrett** [comment]
Yes, it's a much more pleasant approach! I was happy to find it. For context, we're now using targets as our default project structure, but we also use renv by default. Sometimes renv doesn't have a package cached that you otherwise have on your machine, so this workflow is much smoother. We use an explicit, `DESCRIPTION`-based approach to renv package management, and `usethis::use_package()` runs `check_installed()` when adding a pkg to that file. It was such an improvement to my workflow that I thought I just *had* to try to get it in targets for the rest of my workflow 😁 

Thanks also for the example! That's interesting and exactly the sort of use case I think is compelling. 

Anyway, all updated. Let me know if there's anything else

**Comment by wlandau** [comment]
Looks great, thanks again.

FWIW, `targets` has a `tar_renv()` function that detects many of the package requirements of a pipeline and writes a `_packages.R` file that `renv` can detect automatically. (Though it does not create or modify a `DESCRIPTION` file.)

**Comment by malcolmbarrett** [comment]
Awesome, thanks! I didn't know about `tar_renv()`; I'll take a closer look. While it may not be useful for our team workflow, I am developing some teaching material about using targets + renv that uses the default (implicit) renv detection, so it will probably be a great thing to have handy!

**Pull request merged**